src/ps/api/client/mod.rs

```diff
@@ -565,56 +578,119 @@ impl Pennsieve {
 
         into_future_trait(get!(self, "/authentication/cognito-config").and_then(
             move |config_response: serde_json::Value| {
+                let token_pool_value = match config_response.get("tokenPool") {
+                    Some(value) => value,
+                    None => return Either::A(futures::failed(crate::ps::Error::initiate_auth_error(
+                        "Pennsieve server Cognito config missing token pool.",
+                    )))
+                };
+
+                let app_client_id_value = match token_pool_value.get("appClientId") {
+                    Some(value) => value,
+                    None => return Either::A(futures::failed(crate::ps::Error::initiate_auth_error(
+                        "Pennsieve server Cognito config missing token pool client id.",
+                    )))
+                };
+
+                let app_client_id = match app_client_id_value.as_str() {
+                    Some(str) => str,
+                    None => return Either::A(futures::failed(crate::ps::Error::initiate_auth_error(
+                        "Pennsieve server Cognito config token pool client id could not be interpreted as a string.",
+                    )))
+                }.to_string();
+
```

Comment:
This might work to shorten things, but I haven't tried to compile it:

```
into_future_trait(get!(self, "/authentication/cognito-config").then(
  move |config_response: serde_json::Value| {
    let token_pool_value = match config_response.get("tokenPool")
      .ok_or(crate::ps::Error::initiate_auth_error("Pennsieve server Cognito config missing token pool."))?;
    let app_client_id_value = match token_pool_value.get("appClientId")
      .ok_or(crate::ps::Error::initiate_auth_error("Pennsieve server Cognito config missing token pool client id."))?;
    let app_client_id = match app_client_id_value.as_str()
      .map(|client_id| client_id.to_string())
      .ok_or(crate::ps::Error::initiate_auth_error("Pennsieve server Cognito config token pool client id could not be interpreted as a string."))?;
    let request = InitiateAuthRequest {
      analytics_metadata: None,
      auth_flow: "USER_PASSWORD_AUTH".to_string(),
      auth_parameters: Some(auth_parameters),
      client_id: app_client_id,
      client_metadata: None,
      user_context_data: None,
    };
    Ok(request)
  }
)
.then(|request| {
  cognito
     .initiate_auth(request)
     .map_err(Into::into)
     .and_then(move |response| {
       ...
     })
})
```